### PR TITLE
Core Data Concurrency update

### PR DESCRIPTION
**Pull request by melekr**
#### Why?
When in concurrency debugging mode `-com.apple.CoreData.ConcurrencyDebug 1`; any direct call into the context methods from an unexpected thread or queue immediately triggers a concurrency violation.
This requires that a private queue context `.privateQueueConcurrencyType` must be accessed exclusively through its `perform(_:)` or `performAndWait(_:)` methods, which schedule all reads/writes on its private serial queue.

#### Changes:
- Add NSManagedObjectContext extension to capture any thrown error and return the result
- Update public Repository methods to use backgroundContext.performAndWaitThrowing
- Remove direct off-queue fetch/save calls to fix concurrency rule violations
- Move fetch/delete logic into private _locked helpers
- Add doc comments
- Expand Core Data Test coverage

ref: BT-5464

### CODE REVIEW

File: Sources/Features/Extensions/NSManagedObjectContext+Extensions.swift [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                thrownError = error
+            }
+        }
+        if let thrownError = thrownError {
+            throw thrownError
+        }
+        return result
+    }
+    
+    // Swift 5 approach
```
Comment (melekr) [inline Sources/Features/Extensions/NSManagedObjectContext+Extensions.swift:32]: swift 5 approach is cleaner

File: Sources/Features/Extensions/NSManagedObjectContext+Extensions.swift [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    func performAndWaitThrowingSwift5<T>(_ block: () throws -> T) throws -> T {
+        var result: Swift.Result<T, Error>?
+            performAndWait {
+                // Captures returned value or throws error
+                result = Swift.Result(catching: block)
+            }
+            guard let outcome = result else {
+                throw PerformAndWaitError.blockDidNotRun
```
Comment (melekr) [inline Sources/Features/Extensions/NSManagedObjectContext+Extensions.swift:40]: edge case

File: Tests/BacktraceDatabaseTests.swift [hunk 3]
```diff
@@ -51,6 +51,83 @@ final class BacktraceDatabaseTests: QuickSpec {
...
+                                let report = try crashReporter.generateLiveReport(attributes: [:])
+                                try repository.save(report)
+                            } catch {
+                                fail("Failed to save concurrently: \(error)")
+                            }
+                        }
+                    }
+                    
+                    // concurrent reads
+                    for _ in 1...5 {
```
Comment (melekr) [inline Tests/BacktraceDatabaseTests.swift:76]: higher concurrent reads/writes take a lot of time.

**Review by konraddysput — APPROVED** [review]
